### PR TITLE
Observe API

### PR DESCRIPTION
**Pull request by bhelx**
This represents a sneak peek of the future of the Observe API. Up until now, our SDKs have been coupled to our instrumenting compiler. Our goal with Observe is to extend the API to handle all sorts of observability data from your application or existing observability libraries. There will be some follow ups in the near future on this.

Here are the APIs I have implemented in this demo:

* `dylibso_observe.metric(i32, i64, i32)`
* `dylibso_observe.log(i64, i32)`
* `dylibso_observe.span_enter(i64, i32)`
* `dylibso_observe.span_exit()`
* `dylibso_observe.span_tags(i64, i32)`

## Testing

```bash
# with manual instrumentation only
cargo run --example=datadog test/kitchensink.c.wasm
# with manual and automatic instrumentation
cargo run --example=datadog test/kitchensink.c.instr.wasm
```

## `metric(i32, i64, i32)`

This is for allowing the guest to write metrics such as statsd or dogstatsd metrics. The first param is the format (1 for statsd) and it also takes a pointer and length to a string in statsd form. Here is some example C code.

```c
void write_stat(char stat[]) {
  // get our pointers and length
  uintptr_t ptr = (uintptr_t)stat; 
  uint64_t uint64_ptr = (uint64_t)ptr;
  uint32_t uint32_length = (uint32_t)(strlen(stat));
  // call the host function
  metric((uint64_t)1, uint64_ptr, uint32_length);
}

// usage, increment vowels.count metric by 1
write_stat("vowels.count:1|c");
```

## `log(i32, i64, i32)`

This allows you to write log messages, we could also support reading these from stdin and out, but this gives you a direct way to do it. The level maps to rust's log level enum values for error, warn, debug, info, etc.

```c
void write_log(int level, char msg[]) {
  // get our pointers and length
  uintptr_t ptr = (uintptr_t)stat; 
  uint64_t uint64_ptr = (uint64_t)ptr;
  uint32_t uint32_length = (uint32_t)(strlen(msg));
  // call the host function
  statsd((uint64_t)level, uint64_ptr, uint32_length);
}

// usage, write a log at Error level
write_log(1, "An error happened here\n");
```

## `span_enter(i64, i32)` and `span_exit` and `span_tags(i64, i32)`

These APIs allow you to do custom spans. Our normal API functions for the compiler are instrument_enter and instrument_exit. These mostly re-use the implementation of those, so you can use these in an automatically instrumented module if you want. Call span_tags inside of any span to annotate it with tags.

```c
void custom_span_enter(char name[]) {
  uintptr_t ptr = (uintptr_t)name;
  uint64_t uint64_ptr = (uint64_t)ptr;
  uint32_t uint32_length = (uint32_t)(strlen(name));

  span_enter(uint64_ptr, uint32_length);
}

void custom_span_exit() { span_exit(); }

// usage, call span_enter with the name of the function, call the function, then call span_exit.
custom_span_enter("myFunctionName");
myFunctionName();
custom_span_exit();
```

## Rust

In theory, each language could have a small wrapper around these host functions to make them idiomatic to the language. I wrote an example of the rust library. it exposes these host functions but takes normal rust string slices and log level types. It also exposes a macro `instrument` which will automatically instrument the function with the span_enter and exit calls. I've got these crates working but deciding where those guest code should live:

```rust
use log::Level;
use observe_api::{log, span_enter, span_exit, metric, span_tags, MetricFormat};
use observe_instrument::instrument;

#[instrument]
fn log_something(msg: &str) {
    metric(MetricFormat::Statsd, "worlds.helloed:1|c");
    log(Level::Warn, msg);
}

#[instrument]
fn run() {
    log_something("Hello World 1");
    log_something("Hello World 2");
    log_something("Hello World 3");
}

#[instrument]
fn main() {
    span_tags(vec!["user_id:123", "world:hello"]);
    run();
}
```

Screenshot 2023-07-29 at 4 03 49 PM

### CODE REVIEW

**Review by G4Vi — COMMENTED** [review]
So far I only looked at the proposed interfaces. They seem reasonable and i seems like they don't expand scope too much. After working on the wasm wasm-instr-serv, it seems clear that separate pointer and length parameters is the best languages agnostic way to pass strings as doing anything else can require a lot of boilerplate. 

`dylibso_observe.metric(i32, i32, i32)` : Would it be better to have a function to set the metric format or do developers use multiple formats at the same time? I'm not sure if this might be used so often that the overhead of a third parameter is significant. I'm guessing worrying that is premature optimization.

Given that this more the doubles the instrument api, should we consider how we want to handle 64 bit wasm? Just 32 bit support is probably fine for now, but any function that takes a pointer will need 64 bit variants. We could either always use a `i64` for pointers or have separate 32 bit and 64 bit instrument functions?

File: rust/src/adapter/datadog.rs [hunk 1]
```diff
@@ -275,6 +320,35 @@ impl DatadogAdapter {
...
+        //         .open("/tmp/planktonic.log")?;
+
+        //     let mut file_writer = io::BufWriter::new(file);
+        //     for evt in log_events {
+        //         dbg!(&evt);
+        //         file_writer.write_all(evt.message.as_bytes()).unwrap();
+        //     }
+        //     file_writer.flush()?;
+        // }
+
```
Comment (bhelx) [inline rust/src/adapter/datadog.rs:351]: Leaving all this stuff disabled for now. I'll start a new PR with some of the implementation details so we can demo. I didn't want this to break someone in the mean time.

**Comment by bhelx** [comment]
@G4Vi 

> dylibso_observe.metric(i32, i32, i32) : Would it be better to have a function to set the metric format or do developers use multiple formats at the same time? I'm not sure if this might be used so often that the overhead of a third parameter is significant. I'm guessing worrying that is premature optimization.


I think it depends what you mean by overhead. This API will largely be hidden from people. They'll probably be using their normal statsd or prometheus clients and just using a custom dylibso "exporter" or something. It does raise a quesiton about granularity though, might make since to have flavored functions. Like `metric_statsd`, `metric_prometheus`, etc.
I was aiming for lower surface area in terms of number of host functions but open to arguments for a more granular API. Might be worth me writing down a list of what all we'd plan to support in the future.

> Given that this more the doubles the instrument api, should we consider how we want to handle 64 bit wasm? Just 32 bit support is probably fine for now, but any function that takes a pointer will need 64 bit variants. We could either always use a i64 for pointers or have separate 32 bit and 64 bit instrument functions?

Also something I tabled instead of deciding on. Should we just default to 64 bit? What are the consequences to only supporting 64 bit? Does it fallback gracefully in the 32 bit wasm case? I can't recall exactly how it works. I'd prefer to support one API into the future if possible though.

**Comment by G4Vi** [comment]
> I think it depends what you mean by overhead

I just meant the overhead of passing another parameter to the functions. If the metric code for the different format is mostly format dependent, multiple functions sound reasonable, especially if we are supporting <5 formats. 

> What are the consequences to only supporting 64 bit?

Only supporting 64 bit is cumbersome for JS as I believe `BigInt` is needed to pass an i64 param to wasm https://github.com/WebAssembly/JS-BigInt-integration/issues/12. Otherwise, just the potential overhead of the unused 32 bits in most cases. Some care may be needed in the 32 bit pointer to unsigned 64 bit integer conversion to do it all as unsigned to avoid sign extension into the high bits of the 64 bit integer.

**Comment by bhelx** [comment]
Okay, we decided to switch to 64 bit pointers and leave the format parameter in `metric`

File: README.md [hunk 2]
```diff
@@ -40,15 +40,35 @@ prioritize these additional components
...
+You must choose an adapter based on where you want your data to go. At the moment, we support a few systems out of the box.
+In the future we will support a lot more and will have more community driven options. If you don't see support for your
+favorite observability tools feel free to reach out to us at ([support@dylibso.com](mailto:support@dylibso.com)).
+
+Each language includes some examples demonstrating use with different adapters.
+You can view these examples here:
+
+- [Rust](rust/examples)
+- [Go](go/bin)
+- [Js](js/packages)
```
Comment (bhelx) [inline README.md:59]: Is this the right spot?

Comment (nilslice) [reply README.md:59]: we need to do a full clean-up pass over the repo and readme.. i wouldnt let it block you from merging this if you're ready

**Review by nilslice — APPROVED** [review]
nice! very promising start, let's merge and work on some more specific PRs

**Pull request merged**